### PR TITLE
[ENH] Custom hash

### LINKED ISSUE #798: [ENH] Custom hash
Hello,

Is it possible to customize hashing, newer modules becoming more performant like blake 2, 3

Br
Nicolas

### PR DESCRIPTION
**Pull request by NickFillot**
Recover hashing function from os.environ['PYHASH'] like 'package.function'
Change default to blake2b

Closes #798

### CODE REVIEW

File: fsspec/utils.py [hunk 1]
```diff
@@ -4,14 +4,22 @@
...
 
 DEFAULT_BLOCK_SIZE = 5 * 2 ** 20
 PY36 = sys.version_info < (3, 7)
 
+# Hashing: https://www.blake2.net/
+# 64bits perf, should be blake2s for 32bits; need be cross-plateform
+# Waiting for Blake3
+hashing, hasher = os.environ.get('PYHASH', 'hashlib.blake2b').rsplit('.', 1)
+hashing = import_module(hashing)
+hasher = getattr(hashing, hasher)
```
Comment (martindurant) [inline fsspec/utils.py:21]: fsspec already has a function for such string-to-object loading.

File: fsspec/utils.py [hunk 2]
```diff
@@ -4,14 +4,22 @@
...
-from hashlib import md5
 from urllib.parse import urlsplit
 
 DEFAULT_BLOCK_SIZE = 5 * 2 ** 20
 PY36 = sys.version_info < (3, 7)
 
+# Hashing: https://www.blake2.net/
+# 64bits perf, should be blake2s for 32bits; need be cross-plateform
+# Waiting for Blake3
+hashing, hasher = os.environ.get('PYHASH', 'hashlib.blake2b').rsplit('.', 1)
```
Comment (martindurant) [inline fsspec/utils.py:19]: Is this a standard environment variable?
It seems like this should be settable in the fsspec config too, env variable to take precedence if given, but also possible to directly override in code.

File: fsspec/utils.py [hunk 3]
```diff
@@ -314,6 +322,10 @@ def stringify_path(filepath):
         return str(filepath)
     elif hasattr(filepath, "path"):
         return filepath.path
+    elif hasattr(filepath, "name"):
```
[the remark this answers is not in the document]
Comment (NickFillot) [reply fsspec/utils.py:325]: name_function:
- When using glob to write partitioned data (almost all use cases of glob using), partitions with 0, 1, 2 are getting overwritten while need to keep them, which i find weird for a default behavior

stringify_path:
- io.BufferedReader has path in .name
- bytes for usecase

File: fsspec/utils.py [hunk 4]
```diff
@@ -128,28 +136,30 @@ def infer_compression(filename):
 def build_name_function(max_int):
     """Returns a function that receives a single integer
     and returns it as a string padded by enough zero characters
-    to align with maximum possible integer
-
+    to align with maximum possible integer and random uuid
```
[the remark this answers is not in the document]
Comment (NickFillot) [reply fsspec/utils.py:139]: Indeed, i agree with you, but as a default behavior, its almost like spark writing behavior

[the remark this answers is not in the document]
Comment (martindurant) [reply fsspec/utils.py:139]: True, and maybe important as such in some circumstances, but that should depend on the calling code, I think (we have been having a similar discussion for parquet in dask).

**Pull request closed without merging**